File: examples/blink.js
/*
 * example blinks output, blink period is switch with input
 *
 * run:
 * > node blink.js
 *
 * blink.js
 *
 */

"use strict";

const gpiox = require("../gpiox");

const INPUT_PIN = 21;
const OUTPUT_PIN = 20;
const DEBOUNCE_US = 1000; // us
const BLINK_TIME_MS_1 = 500; // ms
const BLINK_TIME_MS_2 = 2500; // ms

console.log("*** blink node.js example ***");
console.log("output pin is gpio", OUTPUT_PIN);
console.log("input pin is gpio", INPUT_PIN, "changes blink period on change");
console.log("stop program with Ctrl+C");

gpiox.init_gpio(OUTPUT_PIN, gpiox.GPIO_MODE_OUTPUT, 0);

gpiox.watch_gpio(INPUT_PIN, gpiox.GPIO_MODE_INPUT_PULLUP, DEBOUNCE_US, gpiox.GPIO_EDGE_BOTH, 
    (state, edge, pin)=>{
        if (state == 0)
            gpiox.blink_gpio(OUTPUT_PIN, BLINK_TIME_MS_2);
        else
            gpiox.blink_gpio(OUTPUT_PIN, BLINK_TIME_MS_1);
    }
);

gpiox.blink_gpio(OUTPUT_PIN, BLINK_TIME_MS_1);

process.on('SIGINT', () => {
    gpiox.deinit_gpio(INPUT_PIN);
    gpiox.deinit_gpio(OUTPUT_PIN);
    console.log(" -> program stopped");
    process.exit();
});

File: cpp/blink.cpp
/*
 * example blinks output, blink period is switch with input
 *
 * build:
 * > make
 *
 * run:
 * > ./blink
 *
 * blink.cpp
 *
 */

#include <stdio.h>
#include <csignal>
#include <cstdlib>

#include "gpiox_lib.h"

#define INPUT_PIN  21
#define OUTPUT_PIN 20

#define DEBOUNCE_US 10000 // us
#define BLINK_TIME_MS_1 500 // ms
#define BLINK_TIME_MS_2 2500 // ms

// signal handler
void onCtrlC(int signum)
{
    gpiox::deinit_gpio(INPUT_PIN);
    gpiox::deinit_gpio(OUTPUT_PIN);

    puts("\nprogram stopped");
    exit(signum);
}

// watch callback
void watch_callback(uint32_t state, uint32_t edge, uint32_t pin)
{
    if (state)
        gpiox::blink_gpio(OUTPUT_PIN, BLINK_TIME_MS_1);
    else
        gpiox::blink_gpio(OUTPUT_PIN, BLINK_TIME_MS_2);
}

int main()
{
    signal(SIGINT, onCtrlC);

    puts("*** blink C++ example ***");
    printf("output pin is gpio%d\n", OUTPUT_PIN);
    printf("input pin gpio%d changes blink period\n", INPUT_PIN);
    puts("stop program with Ctrl+C");

    gpiox::init_gpio(OUTPUT_PIN, gpiox::GPIO_MODE_OUTPUT, 0);
    gpiox::blink_gpio(OUTPUT_PIN, BLINK_TIME_MS_1);

    watch_gpio(INPUT_PIN, gpiox::GPIO_MODE_INPUT_PULLUP, DEBOUNCE_US, gpiox::GPIO_EDGE_BOTH, watch_callback);

    while(1)
    {

    }
}

